web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts

```diff
@@ -0,0 +1,150 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect, useRef } from 'react';
+import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+
+import {
+  ResourcesResponse,
+  ResourceFilter,
+  resourceFilterToHookDeps,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ *
+ * This hook is intended to be used in tandem with the `useInfiniteScroll` hook.
+ * See its documentation for an example.
+ */
+export function useKeyBasedPagination<T>({
+  fetchFunc,
+  clusterId,
+  filter,
+  initialFetchSize = 30,
+  fetchMoreSize = 20,
+}: Props<T>): State<T> {
+  const abortController = useRef<AbortController | null>(null);
+  const { attempt, setAttempt } = useAttempt();
+  const [finished, setFinished] = useState(false);
+  const [resources, setResources] = useState<T[]>([]);
+  const [startKey, setStartKey] = useState(null);
+
+  useEffect(() => {
+    abortController.current?.abort();
+    abortController.current = null;
+
+    setAttempt({ status: '', statusText: '' });
+    setFinished(false);
+    setResources([]);
+    setStartKey(null);
+  }, [clusterId, ...resourceFilterToHookDeps(filter)]);
+
+  // This is an additional countermeasure to prevent fetching if `fetch` is
+  // called multiple times per React state reconciliation cycle.
+  let fetchCalled = false;
+  const fetch = async (force: boolean) => {
+    if (
+      finished ||
+      fetchCalled ||
+      (!force &&
+        (attempt.status === 'processing' || attempt.status === 'failed'))
+    ) {
+      return;
+    }
+
+    try {
+      fetchCalled = true;
+      setAttempt({ status: 'processing' });
+      const limit = resources.length > 0 ? fetchMoreSize : initialFetchSize;
+      abortController.current?.abort();
+      abortController.current = new AbortController();
+      const res = await fetchFunc(
+        clusterId,
+        {
+          ...filter,
+          limit,
+          startKey,
+        },
+        abortController.current.signal
+      );
+      abortController.current = null;
+      setResources([...resources, ...res.agents]);
+      setStartKey(res.startKey);
+      if (!res.startKey) {
+        setFinished(true);
+      }
+      setAttempt({ status: 'success' });
+    } catch (err) {
+      // Aborting is not really an error here.
+      if (isAbortError(err)) {
+        setAttempt({ status: '', statusText: '' });
+        return;
+      }
```

Comment:
Because aborting a request itself shouldn't do anything else: you're not adding anything to the list of resources (they will be reset, but it's a responsibility of a different piece of code); same applies to the finished flag and start key. Aborting a request means "keep the view state as if the request has never happened".